packages/flutter/lib/src/cupertino/segmented_control.dart

```diff
@@ -0,0 +1,568 @@
+            rChildRect = RRect.fromRectAndCorners(childRect, topLeft: const Radius.circular(3.0),
+                bottomLeft: const Radius.circular(3.0));
+          } else if (child == lastChild) {
+            rChildRect = RRect.fromRectAndCorners(childRect, topRight: const Radius.circular(3.0),
+                bottomRight: const Radius.circular(3.0));
+          } else {
+            rChildRect = RRect.fromRectAndCorners(childRect);
+          }
+          _rects.add(rChildRect);
+          start += childWidth;
+          child = childParentData.nextSibling;
+        }
+        break;
+    }
+
+    size = constraints.constrain(overallSize);
+  }
+
+  @override
+  void paint(PaintingContext context, Offset offset) {
+    fillPaint.clear();
+
+    RenderBox child = firstChild;
+    for (T childKey in keys) {
+      final _SegmentedControlContainerBoxParentData childParentData = child.parentData;
+      if (textDirection == TextDirection.ltr) {
+        _paintChild(context, offset, child, childKey, child == firstChild, child == lastChild);
+      } else {
+        _paintChild(context, offset, child, childKey, child == lastChild, child == firstChild);
+      }
+      child = childParentData.nextSibling;
+    }
+  }
+
+  void _paintChild(PaintingContext context, Offset offset, RenderBox child,
+      T childKey, bool isFirst, bool isLast) {
+    if (child == null) {
+      return;
+    }
+
+    final Rect childRect = _boxRect(child).shift(offset);
+    RRect roundedChildRect;
+    if (isFirst) {
+      roundedChildRect = RRect.fromRectAndCorners(childRect, topLeft: const Radius.circular(3.0),
+          bottomLeft: const Radius.circular(3.0));
+    } else if (isLast) {
+      roundedChildRect = RRect.fromRectAndCorners(childRect, topRight: const Radius.circular(3.0),
+          bottomRight: const Radius.circular(3.0));
+    } else {
+      roundedChildRect = RRect.fromRectAndCorners(childRect);
+    }
+
+    Color color = CupertinoColors.white;
+    if (selectedValue != null && selectedValue == childKey) {
+      color = CupertinoColors.activeBlue;
+    } else if (pressedValue != null && pressedValue == childKey) {
+      color = _kPressedBackground;
+    }
+
+    if (textDirection == TextDirection.ltr) {
+      fillPaint.add(new Paint()
+        ..color = color
+        ..style = PaintingStyle.fill);
+    } else {
+      fillPaint.insert(
+          0,
+          new Paint()
+            ..color = color
+            ..style = PaintingStyle.fill);
+    }
+
+    context.canvas.drawRRect(roundedChildRect,
+        textDirection == TextDirection.ltr ? fillPaint.last : fillPaint.first,
+    );
+    context.canvas.drawRRect(
+        roundedChildRect,
+        new Paint()
+          ..color = CupertinoColors.activeBlue
+          ..strokeWidth = 2.0
+          ..style = PaintingStyle.stroke);
+
+    final _SegmentedControlContainerBoxParentData childParentData = child.parentData;
+    context.paintChild(child, childParentData.offset + offset);
+  }
```

Comment:
if you have the layout do everything from left to right, then your entire paint algorithm should be able to be written without having to worry about text direction at all.

Reply:
(or at least could be factored out in the same way so the only part that matters is the walk order)